### PR TITLE
Add metrics host service, for plugins to use

### LINKED ISSUE #827: Metrics inside plugins
Sub-issue of #799

Metrics on core functionality of plugins (eg. `Attest` as in #826) but also on supporting functionality such as `Configure` would also be helpful to get a fuller view into plugins.

### PR DESCRIPTION
**Pull request by amoore877**
**Pull Request check list**

- [X] Commit conforms to CONTRIBUTING.md?
- [X] Proper tests/regressions included?
- [X] Documentation updated?

**Affected functionality**
No existing functionality affected.

**Description of change**
Create new `HostService` to handle metrics. Plugins can onboard to this service in separate PRs.

**Which issue this PR fixes**
fixes #827
Future PRs can add actual usage of the new service into plugins as desired.

### CODE REVIEW

File: pkg/agent/agent.go [hunk 1]
```diff
@@ -50,13 +53,19 @@ func (a *Agent) Run(ctx context.Context) error {
...
+	})
 
 	cat, err := catalog.Load(ctx, catalog.Config{
 		Log: a.c.Log.WithField(telemetry.SubsystemName, telemetry.Catalog),
 		GlobalConfig: catalog.GlobalConfig{
 			TrustDomain: a.c.TrustDomain.Host,
 		},
 		PluginConfig: a.c.PluginConfigs,
+		HostServices: []common_catalog.HostServiceServer{
+			common_services.MetricsServiceHostServiceServer(metricsService),
```
Comment (amoore877) [inline pkg/agent/agent.go]: Note that `Agent` and `Server` have a slightly different pattern on bringing in `HostServices`. `Agent` wants it all declared directly for loading as a slice, whereas `Server` lets the catalog handle slices and just passes in the services themselves.
They should probably be made consistent in another PR.

**Comment by amoore877** [comment]
I'm unsure what the CI errors are; I am not getting `build` script errors locally. What am I missing?

**Comment by azdagron** [comment]
Hmm, I have no idea what is happening with those CI runs...

**Comment by ZymoticB** [comment]
I'll poke it

File: pkg/common/hostservices/metricsservice/metricsservice.go [hunk 2]
```diff
@@ -0,0 +1,68 @@
...
+type Config struct {
+	Metrics telemetry.Metrics
+}
+
+type metricsService struct {
+	cfg Config
+}
+
+// NewMetricsService create and return new Metrics Service
+func NewMetricsService(cfg Config) hostservices.MetricsService {
```
Comment (azdagron) [inline pkg/common/hostservices/metricsservice/metricsservice.go]: nit: rename to `New` to avoid naming stutter

File: pkg/common/hostservices/metricsservice/metricsservice.go [hunk 3]
```diff
@@ -0,0 +1,68 @@
...
+
+func (m metricsService) IncrCounter(ctx context.Context, req *hostservices.IncrCounterRequest) (*hostservices.IncrCounterResponse, error) {
+	labels := convertLabels(req.Labels)
+	m.cfg.Metrics.IncrCounterWithLabels(req.Key, req.Val, labels)
+	return &hostservices.IncrCounterResponse{}, nil
+}
+
+func (m metricsService) MeasureSince(ctx context.Context, req *hostservices.MeasureSinceRequest) (*hostservices.MeasureSinceResponse, error) {
+	labels := convertLabels(req.Labels)
+	m.cfg.Metrics.MeasureSinceWithLabels(req.Key, time.Unix(req.Time, 0), labels)
```
Comment (azdagron) [inline pkg/common/hostservices/metricsservice/metricsservice.go]: We should support sub-second granularity here. go-metrics controls the granularity of what is emitted (defaults to millisecond) but I'd suggest we just use nanoseconds in case that ever becomes configurable.

File: proto/spire/common/hostservices/metricsservice.proto [hunk 4]
```diff
@@ -0,0 +1,60 @@
...
+	float val = 2;
+	repeated Label labels = 3;
+}
+
+message AddSampleResponse {
+}
+
+message MeasureSinceRequest {
+	repeated string key = 1;
+	int64 time = 2;
```
Comment (azdagron) [inline proto/spire/common/hostservices/metricsservice.proto]: we should document the unit this represents

**Comment by ZymoticB** [comment]
@amoore877 could you rebase your change on master? It *may* fix it and travis appears to be queueing jobs right now so I can't debug. Lets try a rebase and I'll come back to this later today if that doesn't resolve.

File: proto/spire/common/hostservices/metricsservice.proto [hunk 5]
```diff
@@ -0,0 +1,60 @@
...
+	repeated Label labels = 3;
+}
+
+message MeasureSinceResponse {
+}
+
+service MetricsService {
+    rpc SetGauge(SetGaugeRequest) returns (SetGaugeResponse);
+		rpc EmitKey(EmitKeyRequest) returns (EmitKeyResponse);
```
Comment (azdagron) [inline proto/spire/common/hostservices/metricsservice.proto]: indentation.... I think we use 4 spaces inside our .proto's

File: pkg/agent/agent.go (hunk 1 above)
Comment (azdagron) [reply pkg/agent/agent.go]: Yep, I agree we should settle on a consistent method here.

Comment (amoore877) [reply pkg/agent/agent.go]: Created #991 to track

File: proto/spire/common/hostservices/metricsservice.proto [hunk 6]
```diff
@@ -0,0 +1,61 @@
...
+    float val = 2;
+    repeated Label labels = 3;
+}
+
+message AddSampleResponse {
+}
+
+message MeasureSinceRequest {
+    repeated string key = 1;
+		// Unix time in nanoseconds
```
Comment (azdagron) [inline proto/spire/common/hostservices/metricsservice.proto]: indentation...

**Comment by azdagron** [comment]
I killed the CI job. I'm working through my own merge queue and will restart the build after updating the branch in due time :)
